src/software/coroutine_main.cpp

```diff
@@ -0,0 +1,510 @@
+    nmo_ptr->execute();
+
+    *data = 3;
+
+    auto fmo_ptr_move = std::move(fmo_ptr);
+    auto dmo_ptr_move = std::move(dmo_ptr);
+    auto nmo_ptr_move = std::move(nmo_ptr);
+
+    fmo_ptr_move->execute();
+    dmo_ptr_move->execute();
+    nmo_ptr_move->execute();
+
+    *data = 4;
+
+    fmo_ptr_move->execute();
+    dmo_ptr_move->execute();
+    nmo_ptr_move->execute();
+
+    // Output
+    // 99
+    // 99
+    // 99
+    // 3
+    // 3
+    // 3
+    // 4
+    // 4
+    // 4
+}
+
+void moveFromStackToHeap() {
+    std::cout << std::endl << std::endl;
+    std::cout << "Test moving from stack to heap" << std::endl;
+
+    auto data = std::make_shared<int>(99);
+
+    FunctionMemberObject fmo(print_data_func, data);
+    DataMemberObject dmo(print_data_func, data);
+    NoMemberObject nmo(print_data_func, data);
+
+    fmo.execute();
+    dmo.execute();
+    nmo.execute();
+
+    *data = 45;
+
+    auto fmo_heap = std::unique_ptr<FunctionMemberObject>();
+    auto dmo_heap = std::unique_ptr<DataMemberObject>();
+    auto nmo_heap = std::unique_ptr<NoMemberObject>();
+
+    *fmo_heap = std::move(fmo);
+    *dmo_heap = std::move(dmo);
+    *nmo_heap = std::move(nmo);
+
+    fmo_heap->execute();
+    dmo_heap->execute();
+    nmo_heap->execute();
+
+    *data = 83;
+
+    fmo_heap->execute();
+    dmo_heap->execute();
+    nmo_heap->execute();
+}
+
+void moveFromHeapToStack() {
+    std::cout << std::endl << std::endl;
+    std::cout << "Test moving from heap to stack" << std::endl;
+
+    auto data = std::make_shared<int>(99);
+
+    auto fmo_heap = std::make_unique<FunctionMemberObject>(print_data_func, data);
+    auto dmo_heap = std::make_unique<DataMemberObject>(print_data_func, data);
+    auto nmo_heap = std::make_unique<NoMemberObject>(print_data_func, data);
+
+    std::optional<FunctionMemberObject> fmo;
+    std::optional<DataMemberObject> dmo;
+    std::optional<NoMemberObject> nmo;
+
+    fmo_heap->execute();
+    dmo_heap->execute();
+    nmo_heap->execute();
+
+    *data = 45;
+
+    *fmo = std::move(*fmo_heap);
+    *dmo = std::move(*dmo_heap);
+    *nmo = std::move(*nmo_heap);
+
+    fmo->execute();
+    dmo->execute();
+    nmo->execute();
+
+    *data = 83;
+
+    fmo->execute();
+    dmo->execute();
+    nmo->execute();
+}
+
+void moveStackToVector()
+{
+    std::cout << std::endl << std::endl;
+    std::cout << "Test move stack variables into vector" << std::endl;
+
+    auto data = std::make_shared<int>(99);
+
+    std::vector<FunctionMemberObject> fmo_vector;
+    fmo_vector.push_back(std::move(FunctionMemberObject(print_data_func, data)));
+    std::vector<DataMemberObject> dmo_vector;
+    dmo_vector.push_back(std::move(DataMemberObject(print_data_func, data)));
+    std::vector<NoMemberObject> nmo_vector;
+    nmo_vector.push_back(std::move(NoMemberObject(print_data_func, data)));
+
+//    fmo_vector.at(0).execute(); // BAD
+//    dmo_vector.at(0).execute(); // BAD
+    nmo_vector.at(0).execute();
```

Comment:
Since each of these are undefined behaviour and they influence each other's error output, it's best if each case were in a different function, e.g. `moveStackToVectorFMO`, `moveStackToVectorDMO`, etc. Similarly for other "BAD" test cases

Reply:
done